lib/PhysicalobjectCsvImporter.class.php

```diff
@@ -0,0 +1,655 @@
+
+      $relation->save($this->dbcon);
+    }
+  }
+
+  /**
+   * Create keymap entry for object
+   *
+   * @param string $sourceName  Name of source data
+   * @param int $sourceId  ID from source data
+   * @param object $object  Object to create entry for
+   *
+   * @return void
+   */
+  public function createKeymapEntry($object, $csvdata)
+  {
+    $keymap = new $this->ormClasses['keymap'];
+    $keymap->sourceName = $this->getOption('sourceName');
+
+    if (!empty($csvdata['legacyId']))
+    {
+      $keymap->sourceId = $csvdata['legacyId'];
+    }
+
+    // Determine target name using object class
+    $keymap->targetName = sfInflector::underscore(
+      str_replace('Qubit', '', get_class($object)));
+    $keymap->targetId   = $object->id;
+
+    $keymap->save();
+  }
+
+  protected function log($msg)
+  {
+    // Just echo to STDOUT for now
+    echo $msg.PHP_EOL;
+  }
+
+  protected function logError($msg)
+  {
+    fwrite($this->getErrorLogHandle(), $msg.PHP_EOL);
+  }
+
+  protected function progressUpdate($data)
+  {
+    $freq = $this->getProgressFrequency();
+
+    if (1 == $freq)
+    {
+      $this->log(sprintf('Imported row [%u/%u]: name "%s"',
+        $this->offset, $this->rowsTotal, $data['name']));
+    }
+    else if ($freq > 1 && 0 == $this->rowsImported % $freq)
+    {
+      $this->log(sprintf('Imported %u of %u rows...',
+        $this->rowsImported, $this->rowsTotal));
+    }
+  }
+
+  protected function getDbConnection()
+  {
+    if (null === $this->dbcon)
+    {
+      $this->dbcon = Propel::getConnection();
+    }
+
+    return $this->dbcon;
+  }
+
+  protected function getErrorLogHandle()
+  {
+    if (null === $filename = $this->getOption('errorLog'))
+    {
+      return STDERR;
+    }
+
+    if (!isset($this->errorLogHandle))
+    {
+      $this->errorLogHandle = fopen($filename, 'w');
+    }
+
+    return $this->errorLogHandle;
+  }
+
+  protected function readCsvFile($filename)
+  {
+    $reader = \League\Csv\Reader::createFromPath($filename, 'r');
+
+    if (!isset($this->options['header']))
+    {
+      // Use first row of CSV file as header
+      $reader->setHeaderOffset(0);
+    }
+
+    $this->rowsTotal = count($reader);
+
+    return $reader;
+  }
+
+  protected function getRecords($stmt)
+  {
+    if (isset($this->options['header']))
+    {
+      $records = $stmt->process($this->reader, $this->options['header']);
+    }
+    else
+    {
+      $records = $stmt->process($this->reader);
+    }
+
+    return $records;
+  }
+
+  protected function processColumn($key, $val, $culture)
+  {
+    switch ($key)
+    {
+      case 'culture':
+        $val = $culture;
+
+        break;
+
+      case 'type':
+        $val = $this->lookupTypeId($val, $culture);
+
+        break;
+
+      case 'descriptionSlugs':
+        $val = $this->processDescriptionSlugs($val);
+
+        break;
+
+      default:
+        $val = trim($val);
+    }
+
+    if ('' !== $val)
```

Comment:
Quibble: could use empty()

Reply:
@sbreker I actually want to return null when `$val = array()` - empty will return true for an empty array.  I'll add a comment in the code to explain the logic.